aten/src/ATen/native/cuda/kernels.cpp

```diff
@@ -0,0 +1,283 @@
+                                          {};
+  void init(const at::Tensor& activation, const at::Tensor& res, const at::Tensor& bias);
+  void prune();
+  void compress();
+  void masked_mm();
+};
+
+
+// https://docs.nvidia.com/cuda/cusparselt/getting_started.html
+// A, B, C, D in the above link corresponds to weight, activation, offset, and output
+
+// does all the initial preparation stuff
+void CusparseLtLinear::init(const at::Tensor& activation, const at::Tensor& res,
+                            const at::Tensor& bias) {
+  int major_cc, minor_cc;
+  CHECK_CUDA(
+      cudaDeviceGetAttribute(&major_cc, cudaDevAttrComputeCapabilityMajor, 0))
+  CHECK_CUDA(
+      cudaDeviceGetAttribute(&minor_cc, cudaDevAttrComputeCapabilityMinor, 0))
+  TORCH_CHECK(((major_cc == 8 && minor_cc == 0) ||
+        (major_cc == 8 && minor_cc == 6)),
+              "cusparseLt is supported only on GPU devices with compute capability == 8.0, 8.6 current: " +
+                    std::to_string(major_cc) + "." + std::to_string(minor_cc));
+
+  // m & k are for weight I think, k & n are for activation
+  // check if weight is transposed?
+  auto m = weight.size(0);
+  auto k = weight.size(1);
+  auto n = activation.size(1); // this is assuming num_batches > 1
+  auto num_batches = activation.size(0);
+  int64_t batch_strideA = m * k;
+  int64_t batch_strideB = k * n;
+  int64_t batch_strideC = m * n;
+  std::cout << "here" << std::endl;
+  // TODO: make these user inputs
+  constexpr auto order = CUSPARSE_ORDER_ROW;
+  constexpr auto type = CUDA_R_16F;
+  constexpr auto compute_type = CUSPARSE_COMPUTE_16F;
+  std::cout << "here1" << std::endl;
+
+  bool is_rowmajor = (order == CUSPARSE_ORDER_ROW);
+  bool isA_transposed = (op_weight != CUSPARSE_OPERATION_NON_TRANSPOSE);
+  bool isB_transposed = (op_activation != CUSPARSE_OPERATION_NON_TRANSPOSE);
+  // TODO: may need to adjust logic if transpose is passed in
+  // TODO: make variable names more descriptive of weight, activation, bias, etc..
+  auto     num_A_rows     = (isA_transposed) ? k : m;
+  auto     num_A_cols     = (isA_transposed) ? m : k;
+  auto     num_B_rows     = (isB_transposed) ? n : k;
+  auto     num_B_cols     = (isB_transposed) ? k : n;
+  auto     num_C_rows     = m;
+  auto     num_C_cols     = n;
+  // is this dtype dependent?
+  unsigned alignment      = 16;
+  auto     lda            = (is_rowmajor) ? num_A_cols : num_A_rows;
+  auto     ldb            = (is_rowmajor) ? num_B_cols : num_B_rows;
+  auto     ldc            = (is_rowmajor) ? num_C_cols : num_C_rows;
+  auto     C_size         = num_batches * batch_strideC;
+  // TODO: make this a function of dtype when dtype is a user input
+  auto     A_size_bytes   = num_batches * batch_strideA * sizeof(c10::Half);
+  auto     B_size_bytes   = num_batches * batch_strideB * sizeof(c10::Half);
+  auto     C_size_bytes   = num_batches * batch_strideC * sizeof(__half);
+  dA = weight.data_ptr<c10::Half>();
```

Comment:
nit: For dtype dispatch you can use https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/Dispatch.h#L210-L225 , but for now half is enough anyway